pkg/cvo/metrics.go

```diff
@@ -105,38 +115,84 @@ type asyncResult struct {
 	error error
 }
 
-// RunMetrics launches an server bound to listenAddress serving
-// Prometheus metrics at /metrics over HTTPS.  Continues serving
-// until runContext.Done() and then attempts a clean shutdown
-// limited by shutdownContext.Done().  Assumes runContext.Done()
-// occurs before or simultaneously with shutdownContext.Done().
-func RunMetrics(runContext context.Context, shutdownContext context.Context, listenAddress string, tlsConfig *tls.Config) error {
-	if tlsConfig == nil {
-		return errors.New("TLS configuration is required to serve metrics")
-	}
-
+func createHttpServer() *http.Server {
 	handler := http.NewServeMux()
 	handler.Handle("/metrics", promhttp.Handler())
 	server := &http.Server{
 		Handler: handler,
 	}
+	return server
+}
 
-	tcpListener, err := net.Listen("tcp", listenAddress)
+func shutdownHttpServer(parentCtx context.Context, svr *http.Server) {
+	ctx, cancel := context.WithTimeout(parentCtx, 5*time.Second)
+	defer cancel()
+	klog.Info("Shutting down metrics server so it can be recreated with updated TLS configuration.")
+	if err := svr.Shutdown(ctx); err != nil {
+		klog.Errorf("Failed to gracefully shut down metrics server during restart: %w", err)
+	}
+}
+
+func startListening(svr *http.Server, tlsConfig *tls.Config, lAddr string, resultChannel chan asyncResult) {
+	tcpListener, err := net.Listen("tcp", lAddr)
 	if err != nil {
-		return err
+		resultChannel <- asyncResult{name: "HTTPS server", error: err}
+	}
+	tlsListener := tls.NewListener(tcpListener, tlsConfig)
```

Comment:
I was wondering why we opened a new listener, instead of attaching a reconfigured server to the existing listener.  But docs for [`Server.Serve`][1]:

> Serve always returns a non-nil error and closes `l`.

so makes sense to me to open a new listener like you are doing here.

Although conveniently, opening the docs helped me notice [`Server.ServeTLS(l net.Listener, certFile, keyFile string)`][2], so perhaps we can drop `makeTLSConfig`.  Which maybe doesn't close the passed in listener?  Anyhow, moving in that direction may be more than we want to bite off in this PR.

[1]: https://pkg.go.dev/net/http#Server.Serve
[2]: https://pkg.go.dev/net/http#Server.ServeTLS

Reply:
> why we opened a new listener

Correct, I tried the former first but `Shutdown` [works by first closing all open listeners](https://cs.opensource.google/go/go/+/go1.17.6:src/net/http/server.go;l=2709).

> more than we want to bite off in this PR.

Agree. It does however appear Shutdown is still used and still closes all open listeners.also close the lis